File: apps/badminton-coach/web/js/core/filters.js
/**
 * Signal conditioning for pose landmarks.
 *
 * Raw MediaPipe landmarks on a player filmed from across a court jitter badly:
 * on the reference clip the raw right-wrist speed peaked at 41 m/s, which is
 * roughly three times the fastest wrist speed ever measured in a badminton
 * smash, so it is noise rather than movement. Everything downstream reads
 * filtered values.
 */

/**
 * A single-channel 1-Euro filter.
 *
 * A plain low-pass filter forces a choice between jitter at rest and lag during
 * a swing, and a smash is over in about 100 ms, so lag is not affordable. The
 * 1-Euro filter (Casiez, Roussel & Vogel, CHI 2012) widens its own cutoff as the
 * signal speeds up, which keeps the racket arm sharp while the standing player
 * stays still.
 */
export class OneEuro {
  /**
   * @param {object} [options]
   * @param {number} [options.minCutoff=1.2] Cutoff in Hz at zero speed: lower is smoother at rest.
   * @param {number} [options.beta=0.25] How fast the cutoff opens with speed: higher means less lag.
   * @param {number} [options.dCutoff=1.0] Cutoff for the internal speed estimate.
   */
  constructor({ minCutoff = 1.2, beta = 0.25, dCutoff = 1.0 } = {}) {
    this.minCutoff = minCutoff;
    this.beta = beta;
    this.dCutoff = dCutoff;
    this.reset();
  }

  reset() {
    this.x = null;
    this.dx = 0;
    this.t = null;
  }

  static alpha(dt, cutoff) {
    const tau = 1 / (2 * Math.PI * cutoff);
    return 1 / (1 + tau / dt);
  }

  /**
   * @param {number} value raw sample
   * @param {number} t sample time in seconds
   * @returns {number} filtered value
   */
  filter(value, t) {
    if (this.x === null || this.t === null || !(t > this.t)) {
      this.x = value;
      this.t = t;
      this.dx = 0;
      return value;
    }
    const dt = t - this.t;
    const dxRaw = (value - this.x) / dt;
    this.dx = this.dx + OneEuro.alpha(dt, this.dCutoff) * (dxRaw - this.dx);
    const cutoff = this.minCutoff + this.beta * Math.abs(this.dx);
    this.x = this.x + OneEuro.alpha(dt, cutoff) * (value - this.x);
    this.t = t;
    return this.x;
  }
}

/** A 1-Euro filter applied independently to every coordinate of a landmark set. */
export class LandmarkFilter {
  /**
   * @param {number} count number of landmarks
   * @param {number} dims coordinates per landmark
   * @param {object} [options] passed to each {@link OneEuro}
   */
  constructor(count, dims = 3, options = {}) {
    this.dims = dims;
    this.channels = Array.from(
      { length: count * dims },
      () => new OneEuro(options),
    );
  }

  reset() {
    for (const c of this.channels) c.reset();
  }

  /**
   * @param {number[][]} points landmarks as arrays of at least `dims` numbers
   * @param {number} t sample time in seconds
   * @returns {number[][]} filtered landmarks; extra columns (visibility) pass through
   */
  filter(points, t) {
    return points.map((p, i) => {
      const out = p.slice();
      for (let d = 0; d < this.dims; d += 1) {
        out[d] = this.channels[i * this.dims + d].filter(p[d], t);
      }
      return out;
    });
  }
}

/**
 * Median over a sliding window.
 *
 * Used for body-scale estimates, where a single bad frame would otherwise
 * rescale every normalised measurement on that frame. On the reference clip the
 * per-frame trunk length wanders between 0.36 m and 0.53 m for a player whose
 * trunk is, of course, one fixed length.
 */
export class RunningMedian {
  /** @param {number} size window length in samples */
  constructor(size = 31) {
    this.size = size;
    this.buffer = [];
  }

  reset() {
    this.buffer = [];
  }

  /** @returns {number} the median including `value`, or NaN if nothing valid yet */
  push(value) {
    if (Number.isFinite(value)) {
      this.buffer.push(value);
      if (this.buffer.length > this.size) this.buffer.shift();
    }
    return this.value;
  }

  get value() {
    if (this.buffer.length === 0) return NaN;
    const sorted = this.buffer.slice().sort((a, b) => a - b);
    const mid = sorted.length >> 1;
    return sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
  }
}

/** Median of an array, ignoring non-finite entries. */
export function median(values) {
  const sorted = values.filter(Number.isFinite).sort((a, b) => a - b);
  if (!sorted.length) return NaN;
  const mid = sorted.length >> 1;
  return sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
}

/** Arithmetic mean, ignoring non-finite entries. */
export function mean(values) {
  const ok = values.filter(Number.isFinite);
  return ok.length ? ok.reduce((a, b) => a + b, 0) / ok.length : NaN;
}

/**
 * Central difference of a scalar series, robust to uneven frame spacing.
 *
 * A central difference has half the noise gain of a forward difference and no
 * half-frame phase shift, which matters because the peak of this derivative is
 * what we call the moment of contact.
 */
export function derivative(values, times) {
  const out = new Array(values.length).fill(NaN);
  for (let i = 1; i < values.length - 1; i += 1) {
    const dt = times[i + 1] - times[i - 1];
    if (dt > 0) out[i] = (values[i + 1] - values[i - 1]) / dt;
  }
  if (values.length > 1) {
    const dt0 = times[1] - times[0];
    if (dt0 > 0) out[0] = (values[1] - values[0]) / dt0;
    const n = values.length - 1;
    const dtn = times[n] - times[n - 1];
    if (dtn > 0) out[n] = (values[n] - values[n - 1]) / dtn;
  }
  return out;
}

File: apps/badminton-coach/analyze/badminton_coach/filters.py
"""Signal conditioning for pose landmarks.

Raw MediaPipe landmarks on a player filmed from across a court jitter badly: on
the reference clip the raw right-wrist speed peaked at 41 m/s, about three times
the fastest wrist speed ever measured in a badminton smash, so it is noise rather
than movement. Everything downstream reads filtered values.

Mirrors ``web/js/core/filters.js``.
"""

from __future__ import annotations

import math
from statistics import median as _median


class OneEuro:
    """A single-channel 1-Euro filter.

    A fixed low-pass filter forces a choice between jitter at rest and lag during
    a swing, and a smash is over in about 100 ms, so lag is not affordable. The
    1-Euro filter (Casiez, Roussel & Vogel, CHI 2012) widens its own cutoff as the
    signal speeds up: still at rest, sharp through the swing.
    """

    def __init__(self, min_cutoff: float = 1.2, beta: float = 0.25, d_cutoff: float = 1.0):
        self.min_cutoff = min_cutoff
        self.beta = beta
        self.d_cutoff = d_cutoff
        self.reset()

    def reset(self) -> None:
        self.x = None
        self.dx = 0.0
        self.t = None

    @staticmethod
    def alpha(dt: float, cutoff: float) -> float:
        tau = 1.0 / (2 * math.pi * cutoff)
        return 1.0 / (1.0 + tau / dt)

    def filter(self, value: float, t: float) -> float:
        if self.x is None or self.t is None or not t > self.t:
            self.x, self.t, self.dx = value, t, 0.0
            return value
        dt = t - self.t
        dx_raw = (value - self.x) / dt
        self.dx += OneEuro.alpha(dt, self.d_cutoff) * (dx_raw - self.dx)
        cutoff = self.min_cutoff + self.beta * abs(self.dx)
        self.x += OneEuro.alpha(dt, cutoff) * (value - self.x)
        self.t = t
        return self.x


class LandmarkFilter:
    """A 1-Euro filter applied independently to every coordinate."""

    def __init__(self, count: int, dims: int = 3, **options):
        self.dims = dims
        self.channels = [OneEuro(**options) for _ in range(count * dims)]

    def reset(self) -> None:
        for c in self.channels:
            c.reset()

    def filter(self, points, t: float) -> list[list[float]]:
        out = []
        for i, p in enumerate(points):
            row = list(p)
            for d in range(self.dims):
                row[d] = self.channels[i * self.dims + d].filter(p[d], t)
            out.append(row)
        return out


class RunningMedian:
    """Median over a sliding window.

    Used for body-scale estimates: one bad frame would otherwise rescale every
    normalised measurement on that frame. On the reference clip the per-frame
    trunk length wanders between 0.36 m and 0.53 m for a trunk that is, of
    course, one fixed length.
    """

    def __init__(self, size: int = 31):
        self.size = size
        self.buffer: list[float] = []

    def reset(self) -> None:
        self.buffer.clear()

    def push(self, value: float) -> float:
        if value is not None and math.isfinite(value):
            self.buffer.append(value)
            if len(self.buffer) > self.size:
                self.buffer.pop(0)
        return self.value

    @property
    def value(self) -> float:
        return _median(self.buffer) if self.buffer else float("nan")


def median(values) -> float:
    ok = [v for v in values if v is not None and math.isfinite(v)]
    return _median(ok) if ok else float("nan")


def mean(values) -> float:
    ok = [v for v in values if v is not None and math.isfinite(v)]
    return sum(ok) / len(ok) if ok else float("nan")


def derivative(values, times) -> list[float]:
    """Central difference of a scalar series, robust to uneven frame spacing.

    Half the noise gain of a forward difference and no half-frame phase shift,
    which matters because the peak of this derivative is what we call contact.
    """
    out = [float("nan")] * len(values)
    for i in range(1, len(values) - 1):
        dt = times[i + 1] - times[i - 1]
        if dt > 0:
            out[i] = (values[i + 1] - values[i - 1]) / dt
    if len(values) > 1:
        dt0 = times[1] - times[0]
        if dt0 > 0:
            out[0] = (values[1] - values[0]) / dt0
        n = len(values) - 1
        dtn = times[n] - times[n - 1]
        if dtn > 0:
            out[n] = (values[n] - values[n - 1]) / dtn
    return out
